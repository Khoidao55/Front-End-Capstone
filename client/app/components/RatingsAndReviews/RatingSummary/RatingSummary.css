.ratingSummary {
  display: flex;
  flex-direction: row;
  background-color: #dfdfdf;
  min-height: 100px;
  margin: 10px;
}

.averageScore {
  margin: 10px 0 0 35px;
  font-size: 70px;
}

.summaryStarRating {
  margin: 25px 0 0 30px;
}

.percentRecommend {
  background-color: #dfdfdf;
  min-height: 20px;
  margin: 0 0 0 45px;
}