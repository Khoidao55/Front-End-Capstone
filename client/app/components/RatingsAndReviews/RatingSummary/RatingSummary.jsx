import React, { useEffect, useState } from 'react';
import propTypes from 'proptypes';
import StarRating from '../../StarRating/StarRating.jsx';
import getAverageRating from '../../helperFunctions/getAverageRating.jsx';
import './RatingSummary.css';

const RatingSummary = ({ metaData }) => {
  const [averageScore, setAverageScore] = useState(0);
  const [percentRecommend, setPercentRecommend] = useState(0);

  useEffect(() => {
    if (metaData.ratings) {
      let didRecommend = 0;
      let didNotRecommend = 0;
      const recommendedArray = Object.entries(metaData.recommended);

      const totalScore = getAverageRating(metaData.ratings);
      setAverageScore(totalScore);

      // find average recommended
      for (let j = 0; j < recommendedArray.length; j += 1) {
        // if the reviewer recommends this product
        if (recommendedArray[j][0] === 'true') {
          // incremement didRecommend
          didRecommend += recommendedArray[j][1] * 1;
        } // else
        if (recommendedArray[j][0] === 'false') {
          // incremement didNotRecommend
          didNotRecommend += recommendedArray[j][1] * 1;
        }
      }
      // get average recommendations and turn into percentage
      // add did and didNot recommend together, divide did by total * 100
      let percentageOfRecommendations = (didRecommend / (didRecommend + didNotRecommend)) * 100;
      percentageOfRecommendations = Math.round(percentageOfRecommendations * 10) / 10;
      setPercentRecommend(percentageOfRecommendations);
    }
  }, [metaData]);

  // conditionally render percent who recommend product if there are 1 or more recommendations
  const NaNCondition = () => {
    if (isNaN(percentRecommend)) {
      return (
        <>
        </>
      );
    }
    return (
      <div className="percentRecommend">
        {percentRecommend}
        % recommend this product
      </div>
    );
  };

  // pass ratings to rating breakdown component
  // pass characteristcs to product breakdown component
  return (
    <>
      <div className="ratingSummary">
        <div className="averageScore">
          {averageScore}
        </div>
        <div className="summaryStarRating">
          <StarRating reviewScore={averageScore || null} setMargin="0 0 0 0" />
        </div>
      </div>
      <NaNCondition />
      <div className="ratingBreakdown">rating breakdown component here</div>
      <div className="productBreakdown">product breakdown component here</div>
    </>
  );
};

RatingSummary.propTypes = {
  metaData: propTypes.object
};

export default RatingSummary;
