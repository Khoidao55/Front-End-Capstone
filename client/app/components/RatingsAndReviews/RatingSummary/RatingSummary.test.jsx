import React from 'react';
import { shallow } from 'enzyme';
import RatingSummary from './RatingSummary';

let metaData = {
  product_id: '13034',
  ratings: {
    3: '1',
    4: '1',
    5: '3',
  },
  recommended: {
    true: '5',
  },
  characteristics: {
    Quality: {
      id: 43655,
      value: '3.4000000000000000',
    },
  },
};

const RemoveAllFiltersRender = () => {
  if (listIsFiltered) {
    return (
      <button type="button" className="remove-review-filters-button" onClick={handleRemoveFilterClick}>Remove all filters</button>
    );
  }
  return (<></>);
};

describe('rating summary: ', () => {
  test('it renders correctly', () => {
    const wrapper = shallow(<RatingSummary metaData={metaData} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('it contains a StarRating component: ', () => {
    const wrapper = shallow(<RatingSummary metaData={metaData} />);

    expect(wrapper.find('StarRating').length).toBe(1);
  });

  test('it contains a recommendations component: ', () => {
    const wrapper = shallow(<RatingSummary metaData={metaData} />);

    expect(wrapper.find('NaNCondition').length).toBe(1);
  });

  test('it contains a RatingBreakdown component: ', () => {
    const wrapper = shallow(<RatingSummary metaData={metaData} />);

    expect(wrapper.find('RatingBreakdown').length).toBe(1);
  });

  test('it contains a RemoveAllFilters component: ', () => {
    const wrapper = shallow(<RatingSummary metaData={metaData} RemoveAllFiltersRender={RemoveAllFiltersRender} />);

    expect(wrapper.find('RemoveAllFiltersRender').length).toBe(1);
  });

  test('it contains a ProductBreakdown component: ', () => {
    const wrapper = shallow(<RatingSummary metaData={metaData} />);

    expect(wrapper.find('ProductBreakdown').length).toBe(1);
  });

});
