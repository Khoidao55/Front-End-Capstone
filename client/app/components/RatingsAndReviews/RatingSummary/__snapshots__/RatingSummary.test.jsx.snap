// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rating summary:  it renders correctly 1`] = `ShallowWrapper {}`;
